svm/src/account_loader.rs

```diff
@@ -212,29 +208,30 @@ fn load_transaction_accounts<CB: TransactionProcessingCallback>(
             let account = if solana_sdk::sysvar::instructions::check_id(key) {
                 construct_instructions_account(message)
             } else {
-                let instruction_account = u8::try_from(i)
-                    .map(|i| instruction_accounts.contains(&&i))
-                    .unwrap_or(false);
                 let (account_size, mut account, rent) = if let Some(account_override) =
                     account_overrides.and_then(|overrides| overrides.get(key))
                 {
                     (account_override.data().len(), account_override.clone(), 0)
-                } else if let Some(program) = (!instruction_account && !message.is_writable(i))
-                    .then_some(())
-                    .and_then(|_| loaded_programs.find(key))
-                {
-                    callbacks
-                        .get_account_shared_data(key)
-                        .ok_or(TransactionError::AccountNotFound)?;
-                    // Optimization to skip loading of accounts which are only used as
-                    // programs in top-level instructions and not passed as instruction accounts.
-                    let program_account = account_shared_data_from_program(&program);
-                    (program.account_size, program_account, 0)
                 } else {
+                    let is_account_writable = message.is_writable(i);
                     callbacks
-                        .get_account_shared_data(key)
-                        .map(|mut account| {
-                            if message.is_writable(i) {
+                        .load_account_with(key, |_account| {
+                            // cache the account in read-only cache
+                            true
```

Comment:
@alessandrod 

https://github.com/anza-xyz/agave/pull/1408#issuecomment-2128383925 is addressed here.